gi/object.cpp

```diff
@@ -2524,6 +2596,189 @@ gjs_add_interface(GType instance_type,
                                 &interface_vtable);
 }
 
+static gboolean
+validate_interfaces_and_properties_args(JSContext *cx,
+                                        JSObject  *interfaces,
+                                        JSObject  *properties,
+                                        guint32   *n_interfaces,
+                                        guint32   *n_properties)
+{
+    guint32 n_int, n_prop;
+
+    if (!JS_IsArrayObject(cx, interfaces)) {
+        gjs_throw(cx, "Invalid parameter interfaces (expected Array)");
+        return FALSE;
+    }
+
+    if (!JS_GetArrayLength(cx, interfaces, &n_int))
+        return FALSE;
+
+    if (!JS_IsArrayObject(cx, properties)) {
+        gjs_throw(cx, "Invalid parameter properties (expected Array)");
+        return FALSE;
+    }
+
+    if (!JS_GetArrayLength(cx, properties, &n_prop))
+        return FALSE;
+
+    if (n_interfaces != NULL)
+        *n_interfaces = n_int;
+    if (n_properties != NULL)
+        *n_properties = n_prop;
+    return TRUE;
+}
+
+static gboolean
+get_interface_gtypes(JSContext *cx,
+                     JSObject  *interfaces,
+                     guint32   n_interfaces,
+                     GType    *iface_types)
+{
+    guint32 i;
+
+    for (i = 0; i < n_interfaces; i++) {
+        jsval iface_val;
+        GType iface_type;
+
+        if (!JS_GetElement(cx, interfaces, i, &iface_val))
+            return FALSE;
+
+        if (!JSVAL_IS_OBJECT(iface_val) ||
+            gjs_throw (cx, "Invalid parameter interfaces (element %d was not a GType)", i);
+            return FALSE;
+        }
+
+        iface_type = gjs_gtype_get_actual_gtype(cx, JSVAL_TO_OBJECT(iface_val));
+        if (iface_type == G_TYPE_INVALID)) {
+            gjs_throw (cx, "Invalid parameter interfaces (element %d was not a GType)", i);
+            return FALSE;
+        }
+
+        iface_types[i] = iface_type;
+    }
+    return TRUE;
+}
+
+static gboolean
+save_properties_for_class_init(JSContext *cx,
+                               JSObject  *properties,
+                               guint32    n_properties,
+                               GType      gtype)
+{
+    GPtrArray *properties_native = NULL;
+    guint32 i;
+    gboolean retval = FALSE;
+
+    if (!class_init_properties)
+        class_init_properties = gjs_hash_table_new_for_gsize((GDestroyNotify) g_ptr_array_unref);
+    properties_native = g_ptr_array_new_with_free_func((GDestroyNotify) g_param_spec_unref);
+    for (i = 0; i < n_properties; i++) {
+        jsval prop_val;
+        JSObject *prop_obj;
+
+        if (!JS_GetElement(cx, properties, i, &prop_val))
+            goto out;
+
+        if (!JSVAL_IS_OBJECT(prop_val)) {
+            gjs_throw(cx, "Invalid parameter, expected object");
+            goto out;
+        }
+        prop_obj = JSVAL_TO_OBJECT(prop_val);
```

Comment:
I should have also mentioned earlier - the new hotness in JSAPI land is to use `JSRootedObject` to get us read for exact stack rooting, so this becomes:

    JSRuntime *runtime = JS_GetRuntime(cx);
    JS::RootedValue prop_val (runtime);

    if (!JS_GetElement(cx, properties, i, &(prop_val.get()))) {
        ... error handling
    }
    JS::RootedObject prop_obj(runtime, prop_val.toObject());

    if (!gjs_typecheck_param(cx, prop_obj, G_TYPE_NONE, JS_TRUE)) {
        ... error handling
    }

Of course, this is all just refactored code, so not a big deal now, just a heads up in case there was other JSAPI stuff I missed.